[UnifiedPDF] [iOS] Support printing in WKWebView

hortont424 opened the pull request:
#### f71c4eb33c8f16ceb5508bc9a414187f93d58203
<pre>
[UnifiedPDF] [iOS] Support printing in WKWebView
<a href="https://bugs.webkit.org/show_bug.cgi?id=287456">https://bugs.webkit.org/show_bug.cgi?id=287456</a>
<a href="https://rdar.apple.com/142292582">rdar://142292582</a>

Reviewed by Abrar Rahman Protyasha.

Hoist the macOS version of PDFPlugin printing into WebPageCocoa and enable it
cross-platform.

The only change made to the code was to move off NSGraphicsContext and hand
PDFKit a CGContextRef directly.

Also, add a test that this works on iOS with Unified PDF enabled.

* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::computePagesForPrintingPDFDocument):
(WebKit::roundCGFloat):
(WebKit::drawPDFPage):
(WebKit::WebPage::drawPDFDocument):
(WebKit::WebPage::drawPagesToPDFFromPDFDocument):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::beginPrinting):
(WebKit::WebPage::drawRectToImage):
(WebKit::WebPage::drawPagesToPDFImpl):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::computePagesForPrintingPDFDocument): Deleted.
(WebKit::WebPage::drawPagesToPDFFromPDFDocument): Deleted.
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::computePagesForPrintingPDFDocument): Deleted.
(WebKit::roundCGFloat): Deleted.
(WebKit::drawPDFPage): Deleted.
(WebKit::WebPage::drawPDFDocument): Deleted.
(WebKit::WebPage::drawPagesToPDFFromPDFDocument): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):

Canonical link: <a href="https://commits.webkit.org/290199@main">https://commits.webkit.org/290199@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69903e611ab3dda0f4a44c0578a89c3f86633cb7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89275 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8799 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94258 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40036 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91326 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9187 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17020 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68781 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26450 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92277 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/7045 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81019 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49142 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6793 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35400 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39143 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36388 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96090 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16455 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16711 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76806 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76955 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18975 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21369 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9575 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16469 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/21780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16210 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19661 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17991 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -1200,6 +1203,115 @@ static bool selectionIsTransparentOrFullyClipped(const VisibleSelection& selecti
+        if (page >= pageCount)
+            break;
+
+        RetainPtr<CFDictionaryRef> pageInfo = adoptCF(CFDictionaryCreateMutable(0, 0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks));
```
Nit: I know you're just moving code around, but CTAD nowadays... 
```suggestion
        RetainPtr pageInfo = adoptCF(CFDictionaryCreateMutable(0, 0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks));
```

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
```diff
@@ -389,6 +396,39 @@ void pressKey(auto key, unsigned short code, Seconds duration = 200_ms)
 
+UNIFIED_PDF_TEST(PrintPDFUsingPrintInteractionController)
+{
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:configurationForWebViewTestingUnifiedPDF().get()]);
```
Nit: explicit smart pointer
```suggestion
    RetainPtr webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:configurationForWebViewTestingUnifiedPDF().get()]);
```

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
```diff
@@ -389,6 +396,39 @@ void pressKey(auto key, unsigned short code, Seconds duration = 200_ms)
+    [webView synchronouslyLoadRequest:request.get()];
+    [webView waitForNextPresentationUpdate];
+
+    auto printPageRenderer = adoptNS([[UIPrintPageRenderer alloc] init]);
```
```suggestion
    RetainPtr printPageRenderer = adoptNS([[UIPrintPageRenderer alloc] init]);
```

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
```diff
@@ -389,6 +396,39 @@ void pressKey(auto key, unsigned short code, Seconds duration = 200_ms)
+    auto printPageRenderer = adoptNS([[UIPrintPageRenderer alloc] init]);
+    [printPageRenderer addPrintFormatter:[webView viewPrintFormatter] startingAtPageAtIndex:0];
+
+    auto printInteractionController = adoptNS([[UIPrintInteractionController alloc] init]);
```
```suggestion
    RetainPtr printInteractionController = adoptNS([[UIPrintInteractionController alloc] init]);
```

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
```diff
@@ -389,6 +396,39 @@ void pressKey(auto key, unsigned short code, Seconds duration = 200_ms)
+    TestWebKitAPI::Util::run(&done);
+    EXPECT_NE([pdfData length], 0UL);
+
+    auto pdf = TestWebKitAPI::TestPDFDocument::createFromData(pdfData.get());
```
```suggestion
    Ref pdf = TestWebKitAPI::TestPDFDocument::createFromData(pdfData.get());
```

hortont424 committed:
[UnifiedPDF] [iOS] Support printing in WKWebView
https://bugs.webkit.org/show_bug.cgi?id=287456
rdar://142292582

Reviewed by Abrar Rahman Protyasha.

Hoist the macOS version of PDFPlugin printing into WebPageCocoa and enable it
cross-platform.

The only change made to the code was to move off NSGraphicsContext and hand
PDFKit a CGContextRef directly.

Also, add a test that this works on iOS with Unified PDF enabled.

* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::computePagesForPrintingPDFDocument):
(WebKit::roundCGFloat):
(WebKit::drawPDFPage):
(WebKit::WebPage::drawPDFDocument):
(WebKit::WebPage::drawPagesToPDFFromPDFDocument):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::beginPrinting):
(WebKit::WebPage::drawRectToImage):
(WebKit::WebPage::drawPagesToPDFImpl):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::computePagesForPrintingPDFDocument): Deleted.
(WebKit::WebPage::drawPagesToPDFFromPDFDocument): Deleted.
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::computePagesForPrintingPDFDocument): Deleted.
(WebKit::roundCGFloat): Deleted.
(WebKit::drawPDFPage): Deleted.
(WebKit::WebPage::drawPDFDocument): Deleted.
(WebKit::WebPage::drawPagesToPDFFromPDFDocument): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):

Canonical link: https://commits.webkit.org/290199@main
